include/libtorrent/disk_io_thread_pool.hpp

```diff
@@ -0,0 +1,133 @@
+/*
+
+Copyright (c) 2005-2016, Arvid Norberg, Steven Siloti
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+* Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+* Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in
+the documentation and/or other materials provided with the distribution.
+* Neither the name of the author nor the names of its
+contributors may be used to endorse or promote products derived
+from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_DISK_IO_THREAD_POOL
+#define TORRENT_DISK_IO_THREAD_POOL
+
+#include "libtorrent/config.hpp"
+#include "libtorrent/export.hpp"
+#include "libtorrent/deadline_timer.hpp"
+#include "libtorrent/io_service_fwd.hpp"
+#include "libtorrent/error_code.hpp"
+
+#include <thread>
+#include <mutex>
+#include <atomic>
+#include <functional>
+
+namespace libtorrent
+{
+	struct pool_thread_interface
```

Comment:
It's handy for testing, but I originally added this interface in anticipation of the future changes outlined in the mmapped I/O document. I expect that we'll end up a separate thread function for at least one of the thread pools. The implementer of this interface is not a thread, as you suggest it's a collection of hooks to connect the thread pool with the worker threads.

More generally I see this interface as providing loose coupling between the thread pool and what the threads themselves do, which in theory might not even be processing a job queue. That abstraction has already leaked a bit with the job_queued function though.